scripts/aem/aem.sh

```diff
@@ -0,0 +1,328 @@
+#!/bin/bash
+
+function error() {
+    echo error: "$@" 1>&2
+    should_die=1
+}
+
+function manage_config() {
+    # create default config if the file doesn't exist
+    if [ ! -f aem-cfg.sh ]; then
+        echo "Creating sample aem-cfg.sh in current directory," \
+             "don't forget to edit it."
+
+        cat > aem-cfg.sh <<'EOF'
+#!/bin/bash
+
+grub_repo=https://github.com/TrenchBoot/grub.git
+xen_repo=https://github.com/3mdeb/xen.git
+
+grub_branch=intel-txt-aem-2.06
+xen_branch=aem/develop
+
+server_prefix=http://10.0.2.2
+server_port=8080
+server_url=$server_prefix:$server_port
+
+boot_disk=/dev/vda
+boot_part=/dev/vda1
+
+# format: {local file under webroot/}:{destination prefix}
+files_to_send=(
+    grub.cfg:/grub
+    bzImage:/grub
+    initramfs.cpio:/grub
+    xen/xen.gz:/grub
+)
+EOF
+    fi
+
+    # source whatever config is there at the moment
+    source aem-cfg.sh
+
+    # do basic sanity checks
+    local vars=(
+        grub_repo xen_repo
+        grub_branch xen_branch
+        server_prefix server_port server_url
+        boot_disk
+        boot_part
+    )
+    for var in "${vars[@]}"; do
+        [ -z "${!var}" ] && error "\$$var is not set in config!"
+    done
+
+    if [[ $boot_part != $boot_disk* ]]; then
+        error "\$boot_disk ($boot_disk) is not a prefix of" \
+              "\$boot_part ($boot_part)!"
+    fi
+
+    if [[ $(declare -p files_to_send) != 'declare -a'* ]]; then
+        error '$files_to_send is not an array.'
+    else
+        for entry in "${files_to_send[@]}"; do
+            IFS=":" read -r file dest_dir <<< "$entry"
+            if [ -z "$file" ] || [ -z "$dest_dir" ]; then
+                error "Wrong \$files_to_send entry: '$entry':"
+                error "  * don't forget : separator"
+                error "  * use / for no destination prefix"
+            fi
+        done
+    fi
+
+    if [ -n "$should_die" ]; then
+        error 'Aborting, please fix "aem-cfg.sh" file in current directory.'
+        exit 1
+    fi
+}
+
+function print_usage() {
+    echo "Usage: $(basename $0) action [action-args...]"
+    echo
+    echo 'Actions:'
+    echo '  build  builds GRUB and/or Xen in docker'
```

Comment:
I think this should go below `init` so actions are listed in more or less chronological order.

Reply:
Oh, I see they are in alphabetical order which is also good.